src/ray/raylet/node_manager.h

```diff
@@ -224,6 +224,25 @@ class NodeManager : public rpc::NodeManagerServiceHandler,
       const std::function<void()> &on_all_replied);
 
  private:
+  template <typename T>
+  void ReleaseWorker(T &&worker_id) {
```

Comment:
Because one caller uses an rvalue and one uses an lvalue so a universal reference will make it work (https://isocpp.org/blog/2012/11/universal-references-in-c11-scott-meyers)

Reply:
An rvalue is a temporary, which means it does not have an address. So, you can't take a reference to it. (Recall that references are basically syntactic sugar for pointers which are addresses.)

In the case of `f(g(x))`, g(x) is an rvalue -- it is not assigned to anything, it is passed in directly to f. In fact, compilers often do an optimization where they directly mov g(x) from eax into the argument of f instead of assigning it to any memory address.

As another example of an rvalue, you could not pass the number 17 into a function that takes `(int& x)` because there is no address for the constant. 

The way that translates to this case, is that the function ReleaseWorker is called on a variable in one case, where you can take a reference, and a temporary in another case, where you can't take a reference. You could assign the temporary to a variable, and then call the function on it, but that's silly.

So, what we do here is called a universal reference (this T&& thing), which will either make a function signature with `&&` or `&` depending on whether it is an rvalue or an lvalue. `auto&&` will also do this, actually, and after C++11 there were lots of improvements for doing this in lambdas.

Now, what is &&? This is called an rvalue reference. It is actually different from the two T&& and auto&& things, which are just special templates. The thing with an rvalue reference is, you can assign rvalues to it. So, `int&& x = 17` works, whereas `int& x = 17` does not compile. Rvalue references don't like lvalues, but you can actually `std::move(var)` into an rvalue reference if you really need to, though in that case it is ownership transfer.

A lot of the motivation for this kind of thing was to support perfect forwarding (no copies), which used to not work in C++. Also, this is kind of a simplification because actually there are xvalues (kind of a combination type), which count as both rvalues and glvalues (generalized lvalues), and prvalues, which are the rvalues that aren't xvalues. The expressions that qualify as each are kind of complicated with a lot of edge cases, but I think it's useful to at least develop a general intuition for the basics.

Reply:
oh i guess my original question was more of where is the caller that is invoking with a temporary here? 

thanks for the detailed explanation anyway, definitely a til!

Reply:
I believe it's `worker->WorkerId()` as it does not return a WorkerId&

Reply:
Ah, just for my own learning, IIUC rvalue could bind to `const &` right? So if the signature is `const WorkerID & worker_id` - does that work for the caller `worker->WorkerId()`?

Reply:
ah, const. Nice!